tracer/src/Datadog.Trace/Iast/DefaultTaintedMap.cs

```diff
@@ -0,0 +1,244 @@
+// <copyright file = "DefaultTaintedMap.cs" company = "Datadog" >
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Datadog.Trace.Iast;
+
+internal class DefaultTaintedMap : ITaintedMap
+{
+    // Default capacity. It MUST be a power of 2.
+    public const int DefaultCapacity = 1 << 14;
+    // Default flat mode threshold.
+    public const int DefaultFlatModeThresold = 1 << 13;
+    // Bitmask to convert hashes to positive integers.
+    public const int PositiveMask = int.MaxValue;
+    // Periodicity of table purges, as number of put operations. It MUST be a power of two.
+    private const int PurgeCount = 1 << 6;
+    // Bitmask for fast modulo with PURGE_COUNT.
+    private const int PurgeMask = PurgeCount - 1;
+    // Map containing the tainted objects
+    private ConcurrentDictionary<int, ITaintedObject> _map;
+    // Bitmask for fast modulo with table length.
+    private int _lengthMask = 0;
+    // Flag to ensure we do not run multiple purges concurrently.
+    private bool _isPurging = false;
+    private object _purgingLock = new();
+    // Number of hash table entries. If the hash table switches to flat mode, it stops counting elements.
+    private int _entriesCount;
+    /* Number of elements in the hash table before switching to flat mode. */
+    private int _flatModeThreshold;
+
+    public DefaultTaintedMap()
+    {
+        _map = new ConcurrentDictionary<int, ITaintedObject>();
```

Comment:
Just wondering if we need the `ITaintedObject` interface, given we always work with `TaintedObject`?